### ISSUE TITLE
`pulumi up/preview -c` changes the config

### ISSUE THREAD

**Issue by JanVoracek** [pulumi/pulumi#3776]
The [docs](https://www.pulumi.com/docs/reference/cli/pulumi_up/#options) say `-c, --config stringArray           Config to use during the update`.

However, it is not used only during the update. As a side effect, it also alters the config file. 

My goal is to pass the current Docker image tag (e.g. `pulumi up -c 'myapp:tag=1d321ef'`) without changing my `Pulumi.<stack>.yaml` as it is tracked by Git. 

Maybe I am missing something? I am migrating from Helm where I had a Helm chart with the configuration in `values.yaml` and then `helm upgrade --set 'myapp.tag=1d321ef' .` did the job.

I've found a similar issue (https://github.com/pulumi/pulumi/issues/2975) but I think environment variables shouldn't be necessary since Pulumi already has its own configuration.

**Comment by gpduck** [comment]
FYI we're also seeing the config files being rewritten when running `pulumi up` on 2.8.0. This is a new behavior introduced between 2.7.1 and 2.8.0.

**Comment by justinvp** [comment]
@gpduck, I think what you're reporting is actually #5126

**Comment by stack72** [comment]
I am going to close this out as a fix for this was released in v2.8.2

**Comment by pawelprazak** [comment]
I can see this behaviour in `v2.20.0`.

When I execute:
```
pulumi up -c redacted:image=gcr.io/redacted/redacted:14bde32
```

A line gets added to:
```
config:
  ...
  redacted:image: gcr.io/redacted/redacted:14bde32
```

Expected behaviour would be to be merged with the config at runtime but not written to the config file.

**Comment by orionstudt** [comment]
Agreed I am also seeing this behavior and am needing a way to pass in an input _for the current run only._

**Comment by justinvp** [comment]
`-c` wasn't intended to be used in this way, it was originally designed for use with `pulumi up <url> --config` for cases where you are deploying code that isn't in the local workspace.

Re-opening this issue to reconsider the design of `-c` to be additional config merged in as part of the `up` operation, but not saved to the config file. This would be a behavior change that we'd likely only do as part of a major version bump.

Related: #2975

**Comment by orionstudt** [comment]
That makes sense historically. I'd say it doesn't need to be a behavioral change of `-c`, it could be a new argument. Just need some way to pass in runtime inputs that isn't environment variables, especially since with automation API environment variables don't really make sense.

**Comment by leezen** [comment]
There's a similar discussion in https://github.com/pulumi/pulumi/issues/3373

**Comment by filip-taxamo** [comment]
It is still happening, also without using -c. 
It rewrites the config each time, getting rid of comments for example.

```sh
infrastructure$ pulumi version
v3.5.1
infrastructure$ git diff -U1 Pulumi.dev.yaml
diff --git a/infrastructure/Pulumi.dev.yaml b/infrastructure/Pulumi.dev.yaml
index c65301d..5c31bb4 100644
--- a/infrastructure/Pulumi.dev.yaml
+++ b/infrastructure/Pulumi.dev.yaml
@@ -4,3 +4,3 @@ config:
   aws:allowedAccountIds:
-  - 1234
+    - 1234  # account-name
   aws:region: eu-west-1
infrastructure$ pulumi preview --diff
Previewing update [...]

Resources:
    133 unchanged
infrastructure$ git diff -U1 Pulumi.dev.yaml
infrastructure$ sudo chattr +i Pulumi.dev.yaml
[sudo] password for filip:
infrastructure$ pulumi preview --diff
error: getting secrets manager: open [...]/infrastructure/Pulumi.dev.yaml: operation not permitted
infrastructure$
```

**Comment by filip-zyzniewski** [comment]
It is still happening, also without using -c. 
It rewrites the config each time, getting rid of comments for example.

```sh
infrastructure$ pulumi version
v3.5.1
infrastructure$ git diff -U1 Pulumi.dev.yaml
diff --git a/infrastructure/Pulumi.dev.yaml b/infrastructure/Pulumi.dev.yaml
index c65301d..5c31bb4 100644
--- a/infrastructure/Pulumi.dev.yaml
+++ b/infrastructure/Pulumi.dev.yaml
@@ -4,3 +4,3 @@ config:
   aws:allowedAccountIds:
-  - 1234
+    - 1234  # account-name
   aws:region: eu-west-1
infrastructure$ pulumi preview --diff
Previewing update [...]

Resources:
    133 unchanged
infrastructure$ git diff -U1 Pulumi.dev.yaml
infrastructure$ sudo chattr +i Pulumi.dev.yaml
[sudo] password for filip:
infrastructure$ pulumi preview --diff
error: getting secrets manager: open [...]/infrastructure/Pulumi.dev.yaml: operation not permitted
infrastructure$
```

**Comment by nickcaballero** [comment]
This would be really helpful. Our team uses Localstack via minikube. Ideally, we could use the FQDN of the Localstack service (localstack.default.svc.cluster.local) but Pulumi doesn't resolve the hostname in macOS (yet, maybe? https://github.com/hashicorp/terraform/issues/3536). So for now we update the configuration in the development environment dynamically, which means that we can't really version control the file.

Once the DNS issue is resolved this would be a non-issue, but in the meantime ad-hoc configuration would allow us to version control the properties that don't change for Localstack:

```
config:
  aws:accessKey: test
  aws:secretKey: test
  aws:region: us-east-1
  aws:skipCredentialsValidation: true
  aws:s3ForcePathStyle: true
  aws:skipMetadataApiCheck: true
  aws:skipRequestingAccountId: true
  aws:skipRegionValidation: true
  aws:skipGetEc2Platforms: true
```

**Comment by karpikpl** [comment]
It's been almost a full year without updates and "-c" still seems to be modifying the config file...

Is there a plan to change this?

**Comment by RobbieMcKinstry** [comment]
This issue isn't currently assigned, but we'd be happy to take a community contribution for a fix. It sounds like `pulumi up -c ...` is probably calling the same code as `pulumi config set ...`. While it's obvious `pulumi config set` should rewrite the config file (because... otherwise, that subcommand would have no effect!), `pulumi up -c` should _not_ rewrite the file.

I'd be willing to mentor a community contribution here. Otherwise, while I can't promise we'll pick this item up from the backlog, I'm hopeful, given that it feels rather straight-forward to fix.

**Comment by Frassle** [comment]
I _really really_ hope no one is relying on -c saving those config settings. Wonder if we should add a warning to -c uses for a few releases before changing it to warn users the behavior will flip.

**Comment by dougludlow** [comment]
I'd be interested in attempting to fix this. Not sure if you're offer to mentor someone through it is still open or not @RobbieMcKinstry?

**Comment by RobbieMcKinstry** [comment]
Hi @dougludlow I'm afraid I'm no longer an OSS maintainer with Pulumi, so I'm sorry to say I no longer have the bandwidth to shepherd new PRs. Thank you very much for offering! Maybe someone else on the core team would be willing to asssist.

**Comment by tgummerer** [comment]
@dougludlow I think we're still happy to mentor a community contribution here.  Let us know if you have any specific questions.

**Comment by tgummerer** [comment]
Actually, re-reading the discussion we had here and some discussion we had internally about this (https://pulumi.slack.com/archives/C02VCJEBT2N/p1710160023943989?thread_ts=1710159495.387689&cid=C02VCJEBT2N as reference for pulumi employees), I don't think we should change the behaviour here, but rather introduce a new flag that doesn't save the config. Something like `--run-config` maybe?

**Comment by orionstudt** [comment]
I was going to suggest the same thing, or like `--config-inline`
